File: vuln-sandbox/python/cwe-916/pbkdf2-low-iterations-safe.py
# SAFE: pbkdf2-low-iterations — PBKDF2 used with sufficient iterations (600,000+)
# Rule: CryptoPbkdf2LowIterations | CWE-916 | Expected: TrueNegative

import hashlib
import os
import binascii


def hash_password(password: str) -> str:
    """Hash a password using PBKDF2-HMAC-SHA256 with 600,000 iterations."""
    salt = os.urandom(32)
    # SAFE: 600,000 iterations meets NIST SP 800-132 recommendation for PBKDF2-HMAC-SHA256
    dk = hashlib.pbkdf2_hmac('sha256', password.encode(), salt, 600000)
    return binascii.hexlify(salt).decode() + ':' + binascii.hexlify(dk).decode()


def verify_password(password: str, stored: str) -> bool:
    """Verify a password against a stored PBKDF2 hash."""
    salt_hex, hash_hex = stored.split(':')
    salt = binascii.unhexlify(salt_hex)
    dk = hashlib.pbkdf2_hmac('sha256', password.encode(), salt, 600000)
    return binascii.hexlify(dk).decode() == hash_hex


if __name__ == '__main__':
    stored = hash_password('my-secure-password')
    print('Verified:', verify_password('my-secure-password', stored))

File: vuln-sandbox/node/cwe-916/pbkdf2-low-iterations-safe.js
// SAFE: pbkdf2-low-iterations — PBKDF2 used with sufficient iterations (600,000+)
// Rule: CryptoPbkdf2LowIterations | CWE-916 | Expected: TrueNegative

const crypto = require('crypto');

function hashPassword(password) {
  const salt = crypto.randomBytes(32);
  // SAFE: 600,000 iterations meets NIST SP 800-132 recommendation for PBKDF2-HMAC-SHA256
  const hash = crypto.pbkdf2Sync(password, salt, 600000, 64, 'sha256');
  return { salt: salt.toString('hex'), hash: hash.toString('hex') };
}

function verifyPassword(password, saltHex, hashHex) {
  const salt = Buffer.from(saltHex, 'hex');
  const hash = crypto.pbkdf2Sync(password, salt, 600000, 64, 'sha256');
  return crypto.timingSafeEqual(hash, Buffer.from(hashHex, 'hex'));
}

const { salt, hash } = hashPassword('my-secure-password');
console.log('Password hashed securely');
console.log('Verification:', verifyPassword('my-secure-password', salt, hash));
